### PR TITLE
Refines handling of 0 and empty height/width attributes

### PR DESCRIPTION
**Pull request by davisshaver**
Alternate branch of https://github.com/Automattic/amp-wp/pull/978#issuecomment-368572107 w/ spacing update.

Fixes https://github.com/Automattic/amp-wp/issues/974

### CODE REVIEW

**Comment by westonruter** [comment]
@davisshaver in #974 you call out numbers with decimals specifically but I don't see any tests that have float dimensions.

**Comment by davisshaver** [comment]
@westonruter See https://github.com/Automattic/amp-wp/pull/979/files#diff-40978c54c34050065a79a2c817f6b0e0R45

File: includes/sanitizers/class-amp-base-sanitizer.php [hunk 1]
```diff
@@ -174,14 +174,25 @@ protected function get_body_node() {
 	 * @return float|int|string Returns a numeric dimension value, or an empty string.
 	 */
 	public function sanitize_dimension( $value, $dimension ) {
-		if ( empty( $value ) ) {
+
+		/**
```
Comment (westonruter) [inline includes/sanitizers/class-amp-base-sanitizer.php]: Minor thing, but this should be just:

```php
// Allows 0 as valid dimension.
```

Same for the comment below.

The `/**` comments are for phpdoc specifically, and there is no need for multi-line here.

File: includes/sanitizers/class-amp-base-sanitizer.php [hunk 2]
```diff
@@ -174,14 +174,25 @@ protected function get_body_node() {
...
 
 		if ( false !== filter_var( $value, FILTER_VALIDATE_INT ) ) {
 			return absint( $value );
 		}
 
+		/**
+		 * Allows floats but parses them to integers.
+		 */
+		if ( false !== filter_var( $value, FILTER_VALIDATE_FLOAT ) ) {
+			return absint( $value );
```
Comment (westonruter) [inline includes/sanitizers/class-amp-base-sanitizer.php]: Should this not be `floor( $value )`?

File: includes/sanitizers/class-amp-img-sanitizer.php [hunk 3]
```diff
@@ -154,18 +154,29 @@ private function determine_dimensions( $need_dimensions ) {
...
 					$node->setAttribute( 'class', $class );
+				} elseif (
+					! is_numeric( $node->getAttribute( 'height' ) ) ) {
+					$height = self::FALLBACK_HEIGHT;
+					$node->setAttribute( 'height', $height );
+					$class = $node->hasAttribute( 'class' ) ? $node->getAttribute( 'class' ) . ' amp-wp-unknown-size amp-wp-unknown-height' : 'amp-wp-unknown-size amp-wp-unknown-height';
+					$node->setAttribute( 'class', $class );
+				} elseif (
+					! is_numeric( $node->getAttribute( 'width' ) )
+				) {
```
Comment (westonruter) [inline includes/sanitizers/class-amp-img-sanitizer.php]: Indentation of this `elseif` line is inconsistent with the previous multi-line `elseif`.

**Comment by davisshaver** [comment]
@westonruter Updated, sry about that.

File: tests/test-amp-img-sanitizer.php [hunk 4]
```diff
@@ -12,42 +12,62 @@ public function setUp() {
...
+
+			'image_with_empty_height'                  => array(
+				'<p><img src="http://placehold.it/300x300" width="300" height="" /></p>',
+				'<p><amp-img src="http://placehold.it/300x300" width="300" height="400" class="amp-wp-unknown-size amp-wp-unknown-height amp-wp-enforced-sizes" sizes="(min-width: 300px) 300px, 100vw"></amp-img></p>',
+			),
+
+			'image_with_zero_width'                    => array(
+				'<p><img src="http://placehold.it/300x300" width="0" height="300" /></p>',
+				'<p><amp-img src="http://placehold.it/300x300" width="0" height="300" sizes="(min-width: 0px) 0px, 100vw" class="amp-wp-enforced-sizes"></amp-img></p>',
+			),
```
Comment (westonruter) [inline tests/test-amp-img-sanitizer.php]: Please also include a test with a zero width _and_ height so we can ensure that the [support topic](https://wordpress.org/support/topic/amp-img-not-setting-height-and-width/) is covered:

```html
<img src="https://example.com/img.jpg" width="0" height="0" border="0">
```

**Comment by davisshaver** [comment]
Thanks for the review, will clean this up tomorrow AM!

File: includes/sanitizers/class-amp-base-sanitizer.php (hunk 2 above)
Comment (davisshaver) [reply includes/sanitizers/class-amp-base-sanitizer.php]: @westonruter Maybe? I am looking at [the docs for floor()](http://php.net/manual/en/function.floor.php) and [round()](http://php.net/manual/en/function.round.php). Both return floats. What about `return absint( round( $value, 0 ) )`?

Comment (westonruter) [reply includes/sanitizers/class-amp-base-sanitizer.php]: `absint( -1.5 )` is `1` but `floor( -1.5 )` is `-1`. In other words, we don't want to lose the sign, do we?

Somewhat strange in the docs there for `floor()`:

> Returns the next lowest integer value (as float)

So I suppose then this is what should be done:

```php
return (int) floor( $value );
```

The `@return` phpdoc should be updated to remove `float`?

Comment (westonruter) [reply includes/sanitizers/class-amp-base-sanitizer.php]: Correction: `floor( -1.5 )` is `2`, as is `floor( -1.1 )`. Maybe `round()` is better then:

```php
return (int) round( $value, 0 );
```

Comment (westonruter) [reply includes/sanitizers/class-amp-base-sanitizer.php]: @davisshaver Hold on. I don't see how decimals are invalid AMP:

image

Comment (davisshaver) [reply includes/sanitizers/class-amp-base-sanitizer.php]: Agree about updating the `@return` statement.

I'd lean towards `absint()` to match the behavior on values that are already `int`, otherwise we may pass a negative height/width.

Comment (westonruter) [reply includes/sanitizers/class-amp-base-sanitizer.php]: True, negative numbers are illegal. But decimal-point values _are_ allowed. So then I think it should be the following with the two conditionals being combined:

```php
if ( is_numeric( $value ) ) {
    return max( 0, floatval( $value ) );
}
```

Comment (davisshaver) [reply includes/sanitizers/class-amp-base-sanitizer.php]: Thanks for the discussion. I will update the PR tomorrow for another consideration

**Comment by davisshaver** [comment]
@westonruter all set for another 👀 ! Much appreciated

File: tests/test-amp-img-sanitizer.php (hunk 4 above)
Comment (westonruter) [reply tests/test-amp-img-sanitizer.php]: Could you add a test for this?

Comment (davisshaver) [reply tests/test-amp-img-sanitizer.php]: Done ✔️

**Comment by sach-patel** [comment]
Found this while searching for a solution where images have fractional value for the height property. It is weird but it happened during migration from Drupal to WP. Is there any way to tackle such images? Thanks.

**Comment by westonruter** [comment]
@sach-patel I wasn't able to see floats causing any validation errors.

**Comment by westonruter** [comment]
@davisshaver I have a related bug if you're interested: https://github.com/Automattic/amp-wp/issues/1030

**Comment by sach-patel** [comment]
@westonruter
Below is an example;
Post: https://thebrag.com/tldr-week-cinema-bad-cop-bad-cop/
AMP version: https://thebrag.com/tldr-week-cinema-bad-cop-bad-cop/amp/

img src="https://i.imgur.com/d1NBb20.gif" alt="" width="425" height="238.85" tag has been converted to amp-anim src="https://i.imgur.com/d1NBb20.gif" alt="" width="425" height="" sizes="(min-width: 425px) 425px, 100vw" class="amp-wp-enforced-sizes"

Google SC is reporting this as "Invalid value for attribute in layout 'height' in tag 'amp-anim'". Is there any quick solution or patch?

Thanks.

**Comment by westonruter** [comment]
@sach-patel the fix is coming in the 0.7 version of the plugin. I suggest testing 0.7-beta2 which should be out this week.

File: includes/sanitizers/class-amp-img-sanitizer.php [hunk 5]
```diff
@@ -154,18 +154,30 @@ private function determine_dimensions( $need_dimensions ) {
 				if ( ! $node instanceof DOMElement ) {
 					continue;
 				}
-
-				// Provide default dimensions for images whose dimensions we couldn't fetch.
-				if ( false !== $dimensions ) {
-					$node->setAttribute( 'width', $dimensions['width'] );
-					$node->setAttribute( 'height', $dimensions['height'] );
```
Comment (westonruter) [inline includes/sanitizers/class-amp-img-sanitizer.php]: These should not have been removed. See #1117.
